harmony_py/auth.py

```diff
@@ -0,0 +1,146 @@
+from requests import Session
+from requests.models import PreparedRequest, Response
+from requests.utils import get_netrc_auth
+from requests_futures.sessions import FuturesSession
+from typing import cast, Optional, Tuple, Union
+
+from .config import Config
+
+
+cfg = Config()
+
+
+def _is_edl_hostname(hostname: str) -> bool:
+    """
+    Determine if a hostname matches an EDL hostname.
+
+    Parameters:
+        hostname (str): A fully-qualified domain name (FQDN).
+
+    Returns:
+        (boolean): True if the hostname is an EDL hostname, else False.
+    """
+    edl_hostname_pattern = r'.*urs\.earthdata\.nasa\.gov$'
+    return re.fullmatch(edl_hostname_pattern, hostname, flags=re.IGNORECASE) is not None
+
+
+class MalformedCredentials(Exception):
+    pass
+
+
+class BadAuthentication(Exception):
+    pass
+
+
+class SessionWithHeaderRedirection(Session):
+    """Modify Authorization headers in accordance with Earthdata Login (EDL) common usage.
+
+    Example:
+        session = SessionWithHeaderRedirection(username, password)
+
+    Parameters:
+        auth (Tuple(str, str)): A tuple of the format ('edl_username', 'edl_password')
+    """
+
+    def __init__(self, auth: Optional[Union[Tuple[str, str], None]] = None) -> None:
+        super().__init__()
+        if auth:
+            self.auth = auth
+        else:
+            self.auth = None
+
+    def rebuild_auth(self, prepared_request: PreparedRequest, response: Response) -> None:
+        """
+        Override Session.rebuild_auth. Strips the Authorization header if neither
+        original URL nor redirected URL belong to an Earthdata Login (EDL) host. Also
+        allows the default requests behavior of searching for relevant .netrc
+        credentials if and only if a username and password weren't provided during
+        object instantiation.
+
+        Parameters:
+            prepared_request (:obj:`PreparedRequest`): Object for the redirection
+            destination.
+            response (:obj:`PreparedRequest`): Object for the where we just came from.
+
+        Returns:
+            (boolean): True if the hostname is an EDL hostname, else False.
+        """
+
+        headers = prepared_request.headers
+        redirect_hostname = cast(str, urlparse(prepared_request.url).hostname)
+        original_hostname = cast(str, urlparse(response.request.url).hostname)
+
+        if 'Authorization' in headers \
+                and (original_hostname != redirect_hostname) \
+                and not _is_edl_hostname(redirect_hostname):
+            del headers['Authorization']
+
+        if self.auth is None:
+            # .netrc might have more auth for us on our new host.
+            new_auth = get_netrc_auth(prepared_request.url) if self.trust_env else None
+            if new_auth is not None:
+                prepared_request.prepare_auth(new_auth)
+
+        return
+
+
+def authenticate(*, auth: Optional[Union[Tuple[str, str], None]] = None,
```

Comment:
(Nit) Given that this method needs to be called by anyone using the library and is the entrypoint to all of the rest of the library:

1. I'd strongly consider moving / mirroring it into __init__.py so that the root package makes it available
2. I'd more weakly consider calling it something like "create_client" (I think we're using the word "session" elsewhere, "create_session" may work as well, though I'm not sure if it implies what we want)

Reply:
Makes sense - I'll plan to do option #1.